diff --git a/perception/autoware_tensorrt_plugins/include/autoware/multi_scale_deform_attn_ops/ms_deform_attn_kernel.hpp b/perception/autoware_tensorrt_plugins/include/autoware/multi_scale_deform_attn_ops/ms_deform_attn_kernel.hpp
--- a/perception/autoware_tensorrt_plugins/include/autoware/multi_scale_deform_attn_ops/ms_deform_attn_kernel.hpp
+++ b/perception/autoware_tensorrt_plugins/include/autoware/multi_scale_deform_attn_ops/ms_deform_attn_kernel.hpp
@@ -42,7 +42,6 @@
 // Copyright (c) OpenMMLab. All rights reserved
 #ifndef AUTOWARE__MULTI_SCALE_DEFORM_ATTN_OPS__MS_DEFORM_ATTN_KERNEL_HPP_
 #define AUTOWARE__MULTI_SCALE_DEFORM_ATTN_OPS__MS_DEFORM_ATTN_KERNEL_HPP_
-#include <cublas_v2.h>
 #include <cuda_runtime.h>
 
 template <typename scalar_t>
